salt/modules/ansiblegate.py

```diff
@@ -0,0 +1,250 @@
+            if os.path.islink(p_el_path):
+                continue
+            if os.path.isdir(p_el_path):
+                paths.update(self._get_modules_map(p_el_path))
+            else:
+                if (any(p_el.startswith(elm) for elm in ['__', '.']) or
+                        not p_el.endswith('.py') or
+                        p_el in ansible.constants.IGNORE_FILES):
+                    continue
+                p_el_path = p_el_path.replace(root, '').split('.')[0]
+                als_name = p_el_path.replace('.', '').replace('/', '', 1).replace('/', '.')
+                paths[als_name] = p_el_path
+
+        return paths
+
+    def load_module(self, module):
+        '''
+        Introspect Ansible module.
+
+        :param module:
+        :return:
+        '''
+        m_ref = self._modules_map.get(module)
+        if m_ref is None:
+            raise LoaderError('Module "{0}" was not found'.format(module))
+        mod = importlib.import_module('ansible.modules{0}'.format(
+            '.'.join([elm.split('.')[0] for elm in m_ref.split(os.path.sep)])))
+
+        return mod
+
+    def get_modules_list(self, pattern=None):
+        '''
+        Return module map references.
+        :return:
+        '''
+        if pattern and '*' not in pattern:
+            pattern = '*{0}*'.format(pattern)
+        modules = []
+        for m_name, m_path in self._modules_map.items():
+            m_path = m_path.split('.')[0]
+            m_name = '.'.join([elm for elm in m_path.split(os.path.sep) if elm])
+            if pattern and fnmatch.fnmatch(m_name, pattern) or not pattern:
+                modules.append(m_name)
+        return sorted(modules)
+
+    def resolve(self):
+        log.debug('Resolving Ansible modules')
+        self._modules_map = self._get_modules_map()
+        return self
+
+    def install(self):
+        log.debug('Installing Ansible modules')
+        return self
+
+
+class AnsibleModuleCaller(object):
+    def __init__(self, resolver):
+        self._resolver = resolver
+
+    def call(self, module, *args, **kwargs):
+        '''
+        Call an Ansible module by invoking it.
+        :param module: the name of the module.
+        :param args: Arguments to the module
+        :param kwargs: keywords to the module
+        :return:
+        '''
+
+        module = self._resolver.load_module(module)
+        if not hasattr(module, 'main'):
+            raise CommandExecutionError('This module is not callable '
+                                        '(see "ansible.help {0}")'.format(module.__name__.replace('ansible.modules.',
+                                                                                                  '')))
+        if args:
+            kwargs['_raw_params'] = ' '.join(args)
+        js_args = '{{"ANSIBLE_MODULE_ARGS": {args}}}'.format(args=json.dumps(kwargs))
+        js_out = subprocess.Popen(["echo", "{0}".format(js_args)], stdout=subprocess.PIPE)
```

Comment:
What do you think about using `salt.utils.timed_subprocess` here and in the line below so that we can bail out if we hang?

Reply:
Good point. Let me check for it. Getting back here ASAP.